theseus/core/objective.py

```diff
@@ -532,11 +544,15 @@ def update_vectorization_if_needed(self):
     def __iter__(self):
         return iter([cf for cf in self.cost_functions.values()])
 
-    def _get_iterator(self):
+    def _get_error_iter_base(self) -> Iterable:
+        return iter(cf for cf in self.cost_functions.values())
+
+    def _get_jacobians_iter(self) -> Iterable:
         self.update_vectorization_if_needed()
-        if self._cost_functions_iterable is None:
-            return iter([cf for cf in self.cost_functions.values()])
-        return iter([cf for cf in self._cost_functions_iterable])
+        if self._jacobians_iter is None:
+            # No vectorization is used, just serve from cost functions
+            return iter(cf for cf in self.cost_functions.values())
+        return iter(cf for cf in self._jacobians_iter)
```

Comment:
Do we need to cache `self._jacobians_iter` here? For example, 
```
if self._jacobians_iter is None:
     self._jacobians_iter = iter(cf for cf in self.cost_functions.values())
return iter(cf for cf in self._jacobians_iter)
```

Reply:
I'm using `None` as a surrogate for `self.vectorized == True`, but maybe this is not ideal. I see if I can replace with 
```python
is self.vectorized:
    return iter(cf for cf in self._jacobians_iter)
else:
    return iter(cf for cf in self.cost_functions.values())
```
I should probably also rename this iterator to `self._vectorized_jacobian_iter`, seems more clear.